feat: main window size

KillerCodeMonkey opened the pull request:
- do not allow to resize main window
- scrollable time entry list
- bottom box with total hours
- label if no time entries found
- move function to format time entry notes for list to lib.ru + tests

frenkel commented:
Works great. Sometimes the entries align exactly at the bottom, making it hard to see that you can scroll the area. How could we make this more obvious? I'm not sure how other programs handle this?

KillerCodeMonkey commented:
dunno, but in general gtk applications are showing a thin dashed line, if there is something to scroll.

KillerCodeMonkey commented:
one possible solution would be to always show the scrollbar, but this is a little bit ugly i think

frenkel commented:
Thunderbird, Firefox and GNOME Terminal make the scrollbar always visible. In which application did you see a dashed line?

KillerCodeMonkey commented:
![Bildschirmfoto vom 2021-02-08 11-17-46](https://user-images.githubusercontent.com/2264672/107206445-55ae5f00-69ff-11eb-818f-e515e05c0b3f.png)
Do you see the light grey dashed line at the end? this thing is visible in the time-tracking app when there is something to scroll.

if you can scroll up it is on top, if down that at the bottom and when you can scroll up and down, it is on top and bottom.

Maybe it is a theme thing on my desktop?

frenkel commented:
Seems like a feature of your distribution or theme indeed, mine (Fedora 33) doesn't show it:

![Schermafdruk van 2021-02-08 12-05-50](https://user-images.githubusercontent.com/121169/107211596-13d4e700-6a06-11eb-8e9a-24d67560bd33.png)

KillerCodeMonkey commented:
@frenkel i changed the scrollbar policies like described in the docs: https://gtk-rs.org/docs/gtk/trait.ScrolledWindowExt.html#tymethod.set_policy 
I changed it for the vertical scrollbar to Always, but then the scrollbar is only shown when the mouse hovers over the scrolled area.

So you can decide what you like more :).

frenkel commented:
Thanks! Which one do you like best? I like this latest version.

KillerCodeMonkey commented:
i do not have any special fav. so feel free to merge :)

frenkel commented:
Is it possible to scroll the view to the bottom when you add a new entry? They now appear outside of the view if you had scrolled upwards previously.

KillerCodeMonkey commented:
i searched for something like this in the ScrolledWindowExt, but i did not find anything. :S

KillerCodeMonkey commented:
@frenkel i implemented the scrolling, but sadly it seems like that the new entry is not in the list (rendered?) when i try to scroll. It is scrolling, but just to the previous last entry. 
Maybe you see the problem?

frenkel commented:
Thanks! I’ll try to take a look next weekend!

frenkel commented on src/ui.rs:
```diff
@@ -350,6 +351,16 @@ impl Ui {
+            Some(scroll) => {
+                if scroll {
+                    let adjustment = self.scroll_view.get_vadjustment().unwrap();
+                    adjustment.set_value(adjustment.get_upper() - adjustment.get_page_size());
```
It appears this code works from within a size adjustment event for grid. Something like this works:

```
self.grid.connect_size_allocate(clone!(scroll_view => move |_,_| {
  let adj = scroll_view.get_vadjustment().unwrap();
  adj.set_value(adj.get_upper() - adj.get_page_size());
}));
```
However, after adding this event handler, any other event will still scroll to bottom, resulting in an unusable edit of the first item for example. Any ideas?

KillerCodeMonkey replied:
thats why i am only try to scroll if creating a new item without using this event handler.

dumb idea from my side: add such a handler on top of the "set_time_entries" method, if a new item was added (scroll_bottom param is true) and at the end we could try to remove this listener.

I will check it

frenkel commented on src/ui.rs:
```diff
@@ -25,7 +25,7 @@ macro_rules! clone {
 pub enum Signal {
     SetTitle(String),
-    SetTimeEntries(Vec<TimeEntry>),
+    SetTimeEntries(Vec<TimeEntry>, Option<bool>),
```
Why not use just `bool` instead of `Option<bool>`?

frenkel commented on src/app.rs:
```diff
@@ -120,7 +120,7 @@ impl App {
     }
 
-    fn retrieve_time_entries(&mut self) {
+    fn retrieve_time_entries(&mut self, scroll_bottom: Option<bool>) {
```
Why not use just `bool` instead of `Option<bool>`?

frenkel commented on src/ui.rs:
```diff
@@ -250,7 +251,7 @@ impl Ui {
     }
 
-    pub fn set_time_entries(&mut self, time_entries: Vec<TimeEntry>) {
+    pub fn set_time_entries(&mut self, time_entries: Vec<TimeEntry>, scroll_bottom: Option<bool>) {
```
Why not use just `bool` instead of `Option<bool>`?

frenkel reviewed (changes requested):
I found the solution, but bumped into another problem, see my comment. Can you get it to work correctly?

frenkel commented:
Your “dumb idea” might work as the only solution 😀

KillerCodeMonkey commented:
but i am really bad at searching through such docs. do you know how to remove an event listener?

KillerCodeMonkey commented:
ah maybe just `disconnect(handlerId)`

frenkel commented:
I think you would need to use something based on `signal::signal_handler_disconnect()`.
Another solution might be to introduce a new `Signal` called "ScrollToBottom" which you would send after the RetrieveTimeEntries signal from the new timer part. That would be more elegant and would call the scroll code just a tiny bit later, which might solve your original bug.

KillerCodeMonkey commented:
after hours of testing and trying... nothing worked. i was not able (maybe because of the lack of rust knowledge) to get it work with the event listener.

first i tried to just globally listen on grid size change and have a scroll_bottom propery on Ui. But i was not able to read the current value of `self.scroll_buttom` inside the closure.

second tried to add a listener per call of `set_time_entries` and remove it inside of the closure... but no luck either.

Then i thought i am ultra clever -> just add a while loop and check, if all children of the grid are `realized`, but it seems everything is realized directly after the `grid.show_all`, which leads to the same strange behavior that it is only scrolled to the second last element.

Seems like everything is updated, but somehow the changing of the dimensions is done later.

So i ended up just adding a little timeout... dumb, but it seems to work.

frenkel commented:
That's too bad, thanks for your testing and investigation! I'll soon have a look myself to try and avoid the timeout.

frenkel commented:
This is a tough one. I couldn't get it to work correctly either. I do want to avoid the timeout though, slow computers might need even more timeout, that will only give trouble. I believe we need to setup the size-allocate signal once and somehow signal (via the global state in UI?) that we do want to scroll or not.

KillerCodeMonkey commented:
I Tried IT already, but i was not able to Pass The boolean (If to scroll or
Not) into The closure :(.

I guess The timeout will Work in every Machine.

If i Understand IT correctly, after The show_all call everything, but The
size allocation is done.


Am 14.02.2021 in 17:42, Frank Groeneveld <notifications@github.com>
schrieb:

This is a tough one. I couldn't get it to work correctly either. I do want
to avoid the timeout though, slow computers might need even more timeout,
that will only give trouble. I believe we need to setup the size-allocate
signal once and somehow signal (via the global state in UI?) that we do
want to scroll or not.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/frenkel/timer-for-harvest/pull/32#issuecomment-778803573>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AARI4YDEQGO6KYBGUARFYW3S674PRANCNFSM4XHVBZNA>
.

frenkel commented:
Last night I tried a number of different things:

- Look in the size-allocate signal at the adjustment and see if you can retrieve the old adjustment. If they differ, we want to scroll. Unfortunately the old adjustment doesn't seem to be available anywhere.
- Look in the changed signal of vadjustment of the scroll_view. This should fire if anything but the value of the vadjustment changes (see https://gtk-rs.org/docs/gtk/trait.AdjustmentExt.html#tymethod.connect_changed) but it appears to fire when you scroll manually as well.

That last option seems to be the nicest way, if we can get it to work correctly. As I understand it, value is your current scroll position. If the total size of the scroll area changes, upper of vadjustment changes and it should fire the signal.

KillerCodeMonkey commented:
so maybe we can store the latest height and if the new height > the one before -> scroll and update the latest height. But there i do not really know again, how i can move the property to the closure of the signal callback to read and write it.

frenkel commented:
Yes, that gets quite complex, using an Arc: https://doc.rust-lang.org/std/sync/struct.Arc.html

frenkel commented:
I've just pushed a working implementation. I'll test this for a few days before merging and releasing a new version.

KillerCodeMonkey commented:
One Thing that Cane to My mind Last night:

Do we Need to handle/avoid scrolling, If the height is Changing when switching days?

frenkel commented:
I thought about that as well. We could probably set the new current height variable to the lower value on switch. I’ll see how it works first the coming days. We’re almost there 😀

frenkel commented:
I'm building a new release now, thank you very much for this great contribution! :tada:

KillerCodeMonkey commented:
YEAAAAHHAAAA!

🚀

Am Fr., 26. Feb. 2021 um 09:22 Uhr schrieb Frank Groeneveld <
notifications@github.com>:

> I'm building a new release now, thank you very much for this great
> contribution! 🎉
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/frenkel/timer-for-harvest/pull/32#issuecomment-786489329>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARI4YCRAIE52ANSW2PT3YDTA5K27ANCNFSM4XHVBZNA>
> .
>

KillerCodeMonkey commented:
@frenkel if i try to install the latest version i get:

require libglib2.0-0 (>= 2.64.6) but 2.64.3-1~ubuntu20.04.1 was installed.
Seems like the requested dependency version does not exist for ubuntu 20.04

frenkel commented:
Same issue as https://github.com/frenkel/timer-for-harvest/issues/33? Try updating first.

KillerCodeMonkey commented:
i tried, but there are no updates for that lib... strange

KillerCodeMonkey commented:
i got it. thought it is part of a stable update, but i have to enable preview updates

KillerCodeMonkey commented:
There it is. the beautiful "minimal" scrollable main window :D
